### PR TITLE
Utopia 551 view pia detail

### PR DESCRIPTION
**Pull request by binggaofw**
# Description

This PR includes the following proposed change(s):

-view pia detail 
Screen Shot 2022-12-20 at 9 11 43 PM

Screen Shot 2022-12-21 at 10 30 15 PM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring / Documentation
- [ ] Version change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Development Dependency Working Agreement
- [x] My code DOES NOT include the importing of new dependencies into the DPIA ecosystem
- [ ] My code DOES include the importing of new dependencies into the DPIA ecosystem
**If new dependencies are being introduced to the DPIA ecosystem:**
- [ ] The functionality of the dependency drastically reduces code complexity and makes my changes more easily maintainable and readible 
- [ ] The dependency being introduced does not contain multiple layers of nested dependencies introducing maintainability complexity to the DPIA ecosystem

## Frontend Development Changes
- [ ] N/A
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to project documentation or diagrams that reflect my changes
- [ ] New and existing unit tests pass locally with my changes
- [ ] My code follows Airbnb React Style Guidelines

## API Development Changes
- [x] N/A
- [ ] I have performed a self-review of my own code
- [ ] My code follows standards and practices outlined in the BC Government API Development Guidelines
- [ ] New and existing unit tests pass locally with my changes
- [ ] My changes includes Swagger documentation updates that reflect the changes I am introducing

## Definition of Done

Definition of Done

### CODE REVIEW

**Comment by Bottle7** [comment]
The details page doesn't look quite right. The area that the information exists on should be much narrower, just like in the Figma designs. I believe that we already have classes which will take care of this for you and also center it on the page.

File: src/frontend/src/pages/PIADetailPage/messages.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+      en: `What are the data or information elements involved in your initiative?`,
+    },
+  },
+  InitiativePISection: {
+    H2Text: {
+      en: `Did you list personal information in question 3?`,
+    },
+  },
+  InitiativeRiskReductionSection: {
+    H2Text: {
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIADetailPage/messages.ts:34]: please group messages wisely. Not sure if it is required here at all.

Also, there are discrepancies in the whole file where a text is not an H2, still called an `H2Text`.  If the message grouping is not required, we can always keep it to just one level as well

**Review by kushal-arora-fw — CHANGES_REQUESTED** [review]

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    (async () => {
+      try {
+        // Actually perform fetch
+        const result = (
+          await HttpRequest.get<IPIAIntakeResponse>(
+            API_ROUTES.GET_PIA_INTAKE.replace(':id', `${id}`),
+          )
+        ).data;
+        setPia(result);
+      } catch (e: any) {
```
[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:33]: In TypeScript, a catch parameter can only be of type any or unknown

**Comment by binggaofw** [comment]
> 

do you know which css class? i can not find it

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        <div>
+          <p className="pb-1">
+            {messages.InitiativeDescriptionSection.H2Text.en}
+          </p>
+
+          <div>
+            <p> {pia.initiativeDescription}</p>
+          </div>
+        </div>
+        <div>
```
[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:128]: that's a mistake, I remove all the space, and there is no space, not sure why add them in first

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:33]: Ohh yeah! We can drop the type on this line and do an `instanceOf Error` check in the code block

**Comment by Bottle7** [comment]
@binggaofw I was thinking of the bcGovPageContainer class and the background class but those don't seem to work. I'm curious as to why the classes we made for creating these layouts aren't working. We should be looking to replicate the structure of the other implementations of this layout.

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:33]: some thing like this ? 
} catch (e) {
        if (e instanceof Error) {
          const errorCode: string = e.cause as string;

**Comment by binggaofw** [comment]
@Bottle7 could you please pull the latest code and test in your local? I think we can resue the background css class

**Comment by Bottle7** [comment]
@binggaofw I will once I finish my current ticket.

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        ).data;
+        setPia(result);
+      } catch (e: any) {
+        const errorCode: string = e.message.split(':')[1];
+        if (errorCode.includes('404')) {
+          win.location = routes.NOT_FOUND;
+        } else if (errorCode.includes('401') || errorCode.includes('403')) {
+          win.location = routes.NOT_AUTHORIZED;
+        } else {
+          throw new Error('Fetch pia failed');
```
[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: Here I am thinking we should add an error alert, what do you think?
```
 } else {
            setFetchPiaError('Something went wrong. Please try again.');
          }
```

[the remark this answers is not in the document]
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: Yeah, we can do that. Please get the copy verified by the Product.

[the remark this answers is not in the document]
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: And the alert can be followed by throwing the error in the console with the details.
`throw new Error('Fetch pia failed');`

File: src/frontend/src/utils/http-request.util.ts [hunk 5]
```diff
@@ -36,7 +36,9 @@ export class HttpRequest {
     const response = await fetch(options.endpoint, config);
 
     if (!response.ok) {
-      throw new Error(`Something went wrong : ${response.status}`);
+      throw new Error('Something went wrong', {
+        cause: new Error(`${response.status}`),
```
Comment (kushal-arora-fw) [inline src/frontend/src/utils/http-request.util.ts:40]:
```suggestion
        cause: response.status,
```

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        const result = (
+          await HttpRequest.get<IPIAIntakeResponse>(
+            API_ROUTES.GET_PIA_INTAKE.replace(':id', `54`),
+          )
+        ).data;
+        setPia(result);
+      } catch (e) {
+        if (e instanceof Error && e.cause instanceof Error) {
+          const errorCode = e.cause.message as string;
+          if (errorCode.includes('404')) {
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIADetailPage/index.tsx:36]: I have added a suggestion below where `cause` is directly status code received, and not an error itself.

The reason I initially suggested this change so we could get rid of string comparisons; and remove `.includes` checks.

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: I have just added a suggestion in your recent commit. See if that looks good

[the remark this answers is not in the document]
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: can not pass the build - error TS2322: Type 'number' is not assignable to type 'Error'.

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 6 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:36]: i get rid of the includes compare and use === instead

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      try {
+        // Actually perform fetch
+        const result = (
+          await HttpRequest.get<IPIAIntakeResponse>(
+            API_ROUTES.GET_PIA_INTAKE.replace(':id', `${id}`),
+          )
+        ).data;
+        setPia(result);
+      } catch (e) {
+        if (e instanceof Error && e.cause instanceof Error) {
```
Comment (antsand) [inline src/frontend/src/pages/PIADetailPage/index.tsx:37]: I don't think this is right. e can be instance of error, e.cause is not necessarily instance of error:

[Description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause#description)

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause
The value of cause can be of any type. You should not make assumptions that the error you caught has an Error as its cause, in the same way that you cannot be sure the variable bound in the catch statement is an Error either. The "Providing structured data as the error cause" example below shows a case where a non-error is deliberately provided as the cause.

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 6 above)
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx]: if we can't simply have `any` type for cause, that's fine. This still will have dependency to the message; But that is fine.

**Review by kushal-arora-fw — COMMENTED** [review]
Good to go form the other changes I suggested. Just pulled the latest code. And I doubt this is matching our designs. Are you done with your css changes?

image

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 7 above)
Comment (kushal-arora-fw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:37]: For the exact same reason, `e.cause` is explicitly created as custom error in the http-request.util.ts;

The idea was to initially have the error status mentioned in the cause, but as Bing found out earlier, it was throwing an error while building, not necessarily at compile time.

check this thread above - https://github.com/bcgov/cirmo-dpia/pull/710#discussion_r1061069621

Comment (antsand) [reply src/frontend/src/pages/PIADetailPage/index.tsx:37]: Ah! Typescript

**Comment by binggaofw** [comment]
@kushal-arora-fw the css issue got fixed, please check again

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+            <div className="row">
+              <div className="col col-md-4">
+                <b>Ministry</b>
+              </div>
+              <div className="col col-md-4">
+                <b>Branch</b>
+              </div>
+            </div>
+            <div className="row">
+              <div className="col col-md-4">{pia.ministry}</div>
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIADetailPage/index.tsx:132]: this should be the name of the ministry, and not the enum returned value from the backend

File: src/frontend/src/pages/PIADetailPage/index.tsx [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        </div>
+        <div className="container pt-5 form__section">
+          <h2 className="pb-3">
+            <b>{messages.GeneralInfoSection.H2TextThree.en}</b>
+          </h2>
+          <div>
+            <p>{messages.InitiativePISection.en}</p>
+
+            <div>
+              <p>{pia.hasAddedPiToDataElements}</p>
```
Comment (kushal-arora-fw) [inline src/frontend/src/pages/PIADetailPage/index.tsx:171]: this should be either of the values used in the dropdown -> Yes, No, I'm not sure

and not true/false/null returned from the backend

**Review by kushal-arora-fw — CHANGES_REQUESTED** [review]
I still see gaps on primarily spacings 🤔 It doesn't align very well. Also, Wondering why are we using MDEditor preview for showing the richtext text.

image

In addition, added a couple of comments too

File: src/frontend/src/pages/PIADetailPage/index.tsx (hunk 8 above)
Comment (binggaofw) [reply src/frontend/src/pages/PIADetailPage/index.tsx:132]: good catch, i will change to full name of ministry

**Comment by binggaofw** [comment]
@kushal-arora-fw I do not find a good way to render text. that's why I am use the markdown editor. if render as text, it will display the original markdown content.

**Comment by binggaofw** [comment]
@kushal-arora-fw  update ministry to full name and PIoption with yes/no/notsure
Screen Shot 2023-01-04 at 1 51 58 PM
Screen Shot 2023-01-04 at 1 52 04 PM

**Comment by kushal-arora-fw** [comment]
> @kushal-arora-fw I do not find a good way to render text. that's why I am use the markdown editor. if render as text, it will display the original markdown content.

I was thinking of react-markdown; However, it might affect custom rich text editor @antsand 's been working on. Anthony - you got any thoughts here aligning to your soon-to-be-in custom editor?

**Comment by binggaofw** [comment]
@kushal-arora-fw I update the css for detail page, now it should align with the form. Please take a look. thanks

**Comment by antsand** [comment]
> 

Depends what the team wants. The editor has been up and working using basic javascript concepts. I have not converted to the react way of doing things, and I may not. (Ran into major issues with reacts and their abstractions)

Most likely something of this sort, I would eventually move it out of DPIA and make it a common library for B.C gov. 

It is seriously up to the team. If they want me to upload the current rich text it is ready. If not, happy to use a library.

**Comment by kushal-arora-fw** [comment]
hmm. The problems if I remember correctly you're referring to are from the class to functional component conversion, right?

**Review by kushal-arora-fw — APPROVED** [review]

**Review by Bottle7 — APPROVED** [review]

**Pull request merged**